File: prototype/rails/src/test/actions/orders/place_order_test.rb
require "test_helper"

module Orders
  class PlaceOrderTest < ActiveSupport::TestCase
    test "it turns the cart into an order the customer can pay for" do
      order = place(create_verified_customer)

      assert_equal Domain::Orders::OrderStatus::AWAITING_PAYMENT, order.status
      assert_equal 45_000, order.subtotal_cents
      assert_equal 45_000, order.total_cents
      assert_equal moment("2026-08-20 09:00:00"), order.placed_at
      assert_nil order.finalized_at
    end

    test "a guest places an order that waits for verification" do
      assert_equal Domain::Orders::OrderStatus::PENDING_VERIFICATION, place(create_anonymous_customer).status
    end

    test "it copies the shipping address onto the order" do
      order = place(create_verified_customer)

      assert_equal "Ada Lovelace", order.shipping_name
      assert_equal "12 Analytical Way", order.shipping_line1
      assert_nil order.shipping_line2
      assert_equal "EC1A 1BB", order.shipping_postal_code
      assert_equal "GB", order.shipping_country
    end

    test "it snapshots the title and price of every item" do
      art = create_listing(title: "Harbour at Dusk", price_cents: 45_000)

      item = order_for(create_verified_customer, art).items.sole

      assert_equal "Harbour at Dusk", item.title
      assert_equal 45_000, item.unit_price_cents
      assert_equal art.seller_id, item.seller_id
    end

    test "it splits the order into one fulfillment per seller" do
      painting = create_listing(create_seller(shop_name: "Blue Kiln Studio"), price_cents: 45_000)
      print = create_listing(create_seller(shop_name: "Rye Press"), price_cents: 10_000)

      order = order_for(create_verified_customer, painting, print)

      assert_equal 55_000, order.subtotal_cents
      assert_equal(
        [[painting.seller_id, 45_000, 4500, 40_500], [print.seller_id, 10_000, 1000, 9000]],
        order.fulfillments.order(:seller_id).map { |f| [f.seller_id, f.subtotal_cents, f.fee_cents, f.net_cents] }
      )
    end

    test "every fulfillment starts awaiting shipment" do
      status = place(create_verified_customer).fulfillments.sole.status

      assert_equal Domain::Orders::FulfillmentStatus::AWAITING_SHIPMENT, status
    end

    test "it takes the stock the order claims" do
      art = create_listing(quantity: 3)
      cart = cart_for(create_verified_customer)
      cart.add(art, quantity: 2, at: moment("2026-08-20 08:00:00"))

      PlaceOrder.new.call(cart: cart, purchaser: purchaser(cart.customer), shipping: shipping_address,
                          now: moment("2026-08-20 09:00:00"))

      art.reload
      assert_equal 1, art.quantity
      assert_equal "for_sale", art.status
    end

    test "the last of a listing marks it sold" do
      art = create_listing(quantity: 1)

      order_for(create_verified_customer, art)

      art.reload
      assert_equal 0, art.quantity
      assert_equal "sold", art.status
    end

    test "it empties the cart" do
      buyer = create_verified_customer
      cart = cart_holding(buyer, create_listing)

      PlaceOrder.new.call(cart: cart, purchaser: purchaser(buyer), shipping: shipping_address,
                          now: moment("2026-08-20 09:00:00"))

      assert_empty cart.reload.items
    end

    test "it refuses an empty cart" do
      buyer = create_verified_customer

      assert_raises(ArgumentError) do
        PlaceOrder.new.call(cart: cart_for(buyer), purchaser: purchaser(buyer), shipping: shipping_address,
                            now: moment("2026-08-20 09:00:00"))
      end
    end

    private

    def place(buyer)
      order_for(buyer, create_listing(price_cents: 45_000))
    end
  end
end

File: prototype/php/src/app/Actions/Orders/PlaceOrderTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Orders;

use App\Actions\Cart\AddToCart;
use App\Domain\DomainRuleViolation;
use App\Domain\Listings\ListingStatus;
use App\Domain\Orders\FulfillmentStatus;
use App\Domain\Orders\OrderStatus;
use App\Models\CustomerBlock;
use App\Models\Order;
use DomainException;

it('turns the cart into an order the customer can pay for', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartWithOneListing($customer, 45000);

    $order = app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($order->status)->toBe(OrderStatus::AwaitingPayment)
        ->and($order->subtotal_cents)->toBe(45000)
        ->and($order->total_cents)->toBe(45000)
        ->and($order->finalized_at)->toBeNull()
        ->and($order->placed_at->format('Y-m-d H:i:s'))->toBe('2026-08-20 09:00:00');
});

it('places an order that waits for verification for an unverified customer', function (): void {
    $customer = $this->anonymousCustomer();
    $cart = $this->cartWithOneListing($customer, 45000);

    $order = app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($order->status)->toBe(OrderStatus::PendingVerification);
});

it('copies the shipping address onto the order', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartWithOneListing($customer, 45000);

    $order = app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($order->shipping_name)->toBe('Ada Lovelace')
        ->and($order->shipping_line1)->toBe('12 Analytical Way')
        ->and($order->shipping_line2)->toBeNull()
        ->and($order->shipping_postal_code)->toBe('EC1A 1BB');
});

it('snapshots the title and price of every item', function (): void {
    $customer = $this->verifiedCustomer();
    $listing = $this->listing($this->seller(), ['title' => 'Harbour at Dusk', 'price_cents' => 45000]);
    $cart = $this->cartFor($customer);
    app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 08:00:00'));

    $order = app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    $item = $order->items()->sole();

    expect($item->title)->toBe('Harbour at Dusk')
        ->and($item->unit_price_cents)->toBe(45000)
        ->and($item->seller_id)->toBe($listing->seller_id);
});

it('splits the order into one fulfillment per seller', function (): void {
    $customer = $this->verifiedCustomer();
    $first = $this->listing($this->seller('Blue Kiln Studio'), ['price_cents' => 45000]);
    $second = $this->listing($this->seller('Rye Press'), ['price_cents' => 10000]);
    $cart = $this->cartFor($customer);
    $addToCart = app(AddToCart::class);
    $addToCart($cart, $first, 1, $this->moment('2026-08-20 08:00:00'));
    $addToCart($cart, $second, 1, $this->moment('2026-08-20 08:00:00'));

    $order = app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($order->subtotal_cents)->toBe(55000);
    expect(
        $order->fulfillments()->orderBy('seller_id')->get()
            ->map(fn ($fulfillment) => [
                $fulfillment->seller_id,
                $fulfillment->subtotal_cents,
                $fulfillment->fee_cents,
                $fulfillment->net_cents,
            ])->all(),
    )->toBe([
        [$first->seller_id, 45000, 4500, 40500],
        [$second->seller_id, 10000, 1000, 9000],
    ]);
});

it('starts every fulfillment awaiting shipment', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartWithOneListing($customer, 45000);

    $order = app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($order->fulfillments()->sole()->status)->toBe(FulfillmentStatus::AwaitingShipment);
});

it('takes the stock the order claims', function (): void {
    $customer = $this->verifiedCustomer();
    $listing = $this->listing($this->seller(), ['price_cents' => 45000, 'quantity' => 3]);
    $cart = $this->cartFor($customer);
    app(AddToCart::class)($cart, $listing, 2, $this->moment('2026-08-20 08:00:00'));

    app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    $listing->refresh();

    expect($listing->quantity)->toBe(1)
        ->and($listing->status)->toBe(ListingStatus::ForSale);
});

it('marks a listing sold when the order claims the last of it', function (): void {
    $customer = $this->verifiedCustomer();
    $listing = $this->listing($this->seller(), ['price_cents' => 45000, 'quantity' => 1]);
    $cart = $this->cartFor($customer);
    app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 08:00:00'));

    app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    $listing->refresh();

    expect($listing->quantity)->toBe(0)
        ->and($listing->status)->toBe(ListingStatus::Sold);
});

it('empties the cart', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartWithOneListing($customer, 45000);

    app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($cart->items()->count())->toBe(0);
});

it('refuses an empty cart', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartFor($customer);

    app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));
})->throws(DomainException::class);

it('refuses a listing that left the storefront while it sat in the cart', function (): void {
    $customer = $this->verifiedCustomer();
    $listing = $this->listing($this->seller(), ['title' => 'Harbour at Dawn', 'price_cents' => 45000]);
    $cart = $this->cartFor($customer);
    app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 08:00:00'));
    $listing->update(['status' => ListingStatus::Archived]);

    $place = fn () => app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($place)->toThrow(DomainRuleViolation::class, '“Harbour at Dawn” is no longer for sale.')
        ->and(Order::count())->toBe(0)
        ->and($cart->items()->count())->toBe(1)
        ->and($listing->refresh()->quantity)->toBe(1)
        ->and($listing->status)->toBe(ListingStatus::Archived);
});

it('refuses a blocked customer', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartWithOneListing($customer, 45000);
    CustomerBlock::factory()->create(['customer_id' => $customer->id, 'reason' => 'Chargeback fraud.']);

    $place = fn () => app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($place)->toThrow(DomainRuleViolation::class, 'Chargeback fraud.')
        ->and(Order::count())->toBe(0);
});

it('refuses a listing whose last unit sold to someone else', function (): void {
    $customer = $this->verifiedCustomer();
    $listing = $this->listing($this->seller(), ['title' => 'Winter Elm', 'price_cents' => 45000, 'quantity' => 1]);
    $cart = $this->cartFor($customer);
    app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 08:00:00'));
    $this->orderFor($this->verifiedCustomer(), $listing);

    $place = fn () => app(PlaceOrder::class)($cart, $this->purchaser($customer), $this->shippingAddress(), $this->moment('2026-08-20 09:00:00'));

    expect($place)->toThrow(DomainRuleViolation::class, '“Winter Elm” is no longer for sale.')
        ->and(Order::count())->toBe(1);
});
